pkg/tempdir/tempdir.go

```diff
@@ -0,0 +1,296 @@
+// If the lock cannot be acquired after MaxInitRetry attempts, it returns an error.
+// The caller should ensure .Cleanup() is called eventually.
+func (td *TempDir) initInstanceTempDir() error {
+	if td.tempDirPath != "" && td.tempDirLock != nil {
+		logrus.Debugf("Temp dir %s already initialized", td.tempDirPath)
+		return nil
+	}
+
+	// Try to create and lock a unique temporary file
+	try := 0
+	for {
+		// Reserve name for lock file in the RootDir.
+		tempLockFile, err := os.CreateTemp(td.RootDir, TempdirLockPrefix)
+		if err != nil {
+			return fmt.Errorf("creating temp directory in %s failed: %w", td.RootDir, err)
+		}
+		tempLockFile.Close()
+
+		tempDirLockPath := tempLockFile.Name()
+		tempDirLock, err := lockfile.GetLockFile(tempDirLockPath)
+		if err != nil {
+			return fmt.Errorf("getting instance lock file %s for temp dir: %w", tempDirLockPath, err)
+		}
+		if err := tempDirLock.TryLock(); err != nil {
+			if try++; try < MaxInitRetry {
+				continue
+			}
+			return fmt.Errorf("locking instance lock for tempdir failed: %w", err)
+		}
+		td.tempDirLock = tempDirLock
+		td.tempDirLockPath = tempDirLockPath
+		break
+	}
+
+	// Create the temporary directory that corresponds to the lock file
+	id := strings.TrimPrefix(filepath.Base(td.tempDirLockPath), TempdirLockPrefix)
+	actualTempDirPath := filepath.Join(td.RootDir, TempDirPrefix+id)
+	if err := os.MkdirAll(actualTempDirPath, 0o700); err != nil {
+		return fmt.Errorf("creating temp directory %s failed: %w", actualTempDirPath, err)
+	}
+
+	td.tempDirPath = actualTempDirPath
+	td.counter = 0
+	return nil
+}
+
+// NewTempDir prepares a TempDir manager for a temporary directory in the specified RootDir.
+// The RootDir itself will be created if it doesn't exist.
+// It performs recovery of stale temporary directories within RootDir (removing those not in active use).
+// The actual instance-specific temporary directory is created lazily on the first call to Add().
+func NewTempDir(rootDir string) (*TempDir, error) {
+	if err := os.MkdirAll(rootDir, 0o700); err != nil {
+		return nil, fmt.Errorf("creating root temp directory %s failed: %w", rootDir, err)
+	}
+
+	td := &TempDir{
+		RootDir: rootDir,
+	}
+
+	if err := RecoverStaleDirs(rootDir); err != nil {
+		logrus.Warnf("Error during stale temp dir recovery in %s: %v", rootDir, err)
+	}
+
+	return td, nil
+}
+
+// Add moves the specified file into the instance's temporary directory.
+// If the instance's temporary directory doesn't exist yet, Add will create and lock it.
+// Files are renamed with a counter-based prefix (e.g., "0-filename", "1-filename") to ensure uniqueness.
+// Note: 'path' must be on the same filesystem as the TempDir for os.Rename to work.
+// The caller MUST ensure .Cleanup() is called.
+func (td *TempDir) Add(path string) error {
+	if td.tempDirLock == nil {
+		if err := td.initInstanceTempDir(); err != nil {
+			return fmt.Errorf("initializing instance temp dir failed: %w", err)
+		}
+	}
```

Comment:
(If `NewTempDir` is called only by code that intends to delete something, this initialization can happen in there, as an ~ordinary constructor. But let’s settle the API shape first.)